src/librustc/middle/region.rs

```diff
@@ -246,8 +246,36 @@ pub struct ScopeTree {
     closure_tree: FxHashMap<hir::ItemLocalId, hir::ItemLocalId>,
 
     /// If there are any `yield` nested within a scope, this map
-    /// stores the `Span` of the last one and the number of expressions
-    /// which came before it in a generator body.
+    /// stores the `Span` of the last one and its index in the
+    /// postorder of the Visitor traversal on the HIR.
+    ///
+    /// HIR Visitor postorder indexes might seem like a peculiar
+    /// thing to care about. but it turns out that HIR bindings
+    /// and the temporary results of HIR expressions are never
+    /// storage-live at the end of HIR nodes with postorder indexes
+    /// lower than theirs, and therefore don't need to be suspended
+    /// at yield-points at these indexes.
+    ///
+    /// Let's show that: let `D` be our binding/temporary and `U` be our
+    /// other HIR node, with `HIR-postorder(U) < HIR-postorder(D)`.
+    ///
+    /// Then:
+    ///     1. From the ordering guarantee of HIR visitors (see
+    ///     `rustc::hir::intravisit`), `D` does not dominate `U`.
```

Comment:
All we know is that U comes before D in the postorder, therefore we know that U may end before D ends in execution order. From this we are concluding that *End(D)* does not dominate *End(U)*, right? That makes sense.

There is a kind of unwritten assumption that the value produced by D is not produced until End(D), I think, right? (Which makes sense.)

Reply:
> There is a kind of unwritten assumption that the value produced by `D` is not produced until `End(D)`, I think, right? (Which makes sense.)

Yeah.